modules/extensions/src/terrain/terrain-effect.ts

```diff
@@ -0,0 +1,238 @@
+      this.initialize(gl);
+      for (const layer of opts.layers) {
+        // Force the terrain layer (and its descendents) to rebuild their models with the new shader
+        if (layer.props.operation.includes('terrain')) {
+          layer.setChangeFlags({extensionsChanged: true});
+        }
+      }
+    }
+
+    const {viewports, pass} = opts;
+    const isPicking = pass.startsWith('picking');
+    this.isPicking = isPicking;
+
+    // TODO - support multiple views?
+    const viewport = viewports[0];
+    const layers = (isPicking ? this.terrainPickingPass : this.terrainPass).getRenderableLayers(
+      viewport,
+      opts as TerrainPickingPassRenderOptions
+    );
+
+    const terrainLayers = layers.filter(l => l.props.operation.includes('terrain'));
+    if (terrainLayers.length === 0) {
+      return;
+    }
+
+    if (!isPicking) {
+      const offsetLayers = layers.filter(l => l.state.terrainFittingMode === 'offset');
+      if (offsetLayers.length > 0) {
+        this._renderHeightMap(gl, terrainLayers, viewport, opts);
+      }
+    }
+
+    const drapeLayers = layers.filter(l => l.state.terrainFittingMode === 'drape');
+    this._updateTerrainCovers(terrainLayers, drapeLayers, viewport, opts);
+  }
+
+  getModuleParameters(layer: Layer): TerrainModuleSettings {
+    const terrainCover = this.terrainCovers.get(layer.id);
+    const {terrainFittingMode} = layer.state;
+
+    let terrainCoverTexture: Texture2D | null;
+    if (!terrainCover) {
+      terrainCoverTexture = null;
+    } else if (this.isPicking) {
+      terrainCoverTexture = terrainCover.getPickingFramebuffer();
+    } else {
+      terrainCoverTexture = terrainCover.getRenderFramebuffer();
+    }
+
+    return {
+      heightMap: this.heightMap?.color,
+      dummyHeightMap: this.dummyHeightMap,
+      terrainCover: terrainCoverTexture,
+      terrainCoverBounds: terrainCover?.commonBounds,
+      useTerrainHeightMap: terrainFittingMode === 'offset',
+      terrainSkipRender: terrainFittingMode === 'drape' || !layer.props.operation.includes('draw')
+    };
+  }
+
+  cleanup(): void {
+    if (this.dummyHeightMap) {
+      this.dummyHeightMap.delete();
+      this.dummyHeightMap = undefined;
+    }
+
+    if (this.heightMap) {
+      this.heightMap.color.delete();
+      this.heightMap.delete();
+      this.heightMap = undefined;
+    }
+
+    for (const terrainCover of this.terrainCovers.values()) {
+      terrainCover.delete();
+    }
+    this.terrainCovers.clear();
+  }
+
+  private _renderHeightMap(
+    gl: WebGLRenderingContext,
+    terrainLayers: Layer[],
+    viewport: Viewport,
+    opts: PreRenderOptions
+  ) {
+    if (!this.heightMap) {
+      // Not supported
+      return;
+    }
+
+    this.heightMap.resize(viewport);
+
+    this.terrainPass.renderHeightMap(this.heightMap, {
+      ...opts,
+      pass: 'terrain-height-map',
+      viewports: [viewport],
```

Comment:
As with the `TerrainCover` we should enlarge the viewport for the heightmap rendering to capture features which are offset by terrain outside the visible viewport

https://user-images.githubusercontent.com/453755/216320779-0e5862c5-dff3-4841-bcbf-a61818c0cdf6.mov

Reply:
Right now there's no way for us to tell the 3D boundary of an arbitrary layer. E.g. ColumnLayer's elevation is calculated as `instancePosition.z + instanceElevation * elevationScale`.